### PR TITLE
clean up profiles

### PR DESCRIPTION
**Pull request by cqc-melf**
# Description

Clean up and first steps of adding support for the base profile

This PR is moving all the code used by all profiles from the two QirGenerator classes to an abstract class.

Adding support for `ADAPTIVE_CREGSIZE` option which is generating profile compatible QIR with assuring the registers size of the integers representing them. This is not yet working on the hardware, so I have not added that to the changelog.

I will add the changes for the base profile in a second PR.

It might be easier to review the commits, they are structures to keep the renaming and changes in the QIR files separated from the code changes.

# Related issues

Solves https://github.com/CQCL/pytket-qir/issues/168

# Checklist

- [x] I have performed a self-review of my code.
- [ ] I have commented hard-to-understand parts of my code.
- [ ] I have made corresponding changes to the public API documentation.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have updated the changelog with any user-facing changes.

### CODE REVIEW

File: pytket/qir/conversion/api.py [hunk 1]
```diff
@@ -60,11 +69,17 @@ def pytket_to_qir(
...
-        You can find more details about the adaptive profile under:
-        https://github.com/qir-alliance/qir-spec/pull/35
-        and soon at:
+    :param profile: generates QIR corresponding to the selected profile:
+        Use QIRProfile.BASE for the base profile, see:
+        https://github.com/qir-alliance/qir-spec/blob/main/specification/under_development/profiles/Base_Profile.md
+        Use QIRProfile.ADAPTIVE for the adaptive profile, see:
         https://github.com/qir-alliance/qir-spec/tree/main/specification/under_development/profiles/Adaptive_Profile.md
+        Use QIRProfile.ADAPTIVE_CREGSIZE for the adaptive profile with additional
+        truncation operation to assure that intergers matching the classical
```
Comment (cqc-alec) [inline pytket/qir/conversion/api.py]:
```suggestion
        truncation operation to assure that integers matching the classical
```

File: pytket/qir/conversion/qirgenerator.py [hunk 2]
```diff
@@ -112,7 +112,7 @@
 }
 
 
-class QirGenerator:
+class AbsQirGenerator:
```
Comment (cqc-alec) [inline pytket/qir/conversion/qirgenerator.py]: Why this name?

File: pytket/qir/conversion/pytketqirgenerator.py [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This module contains all functionality to generate QIR files
+from pytket circuits.
+"""
```
Comment (cqc-alec) [inline pytket/qir/conversion/pytketqirgenerator.py]: Replace these "module-level" descriptions with "file-level" descriptions?

File: pytket/qir/conversion/profileqirgenerator.py [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+    OpType,
+)
+
+from .module import tketqirModule
+from .qirgenerator import (
+    AbsQirGenerator,
+)
+
+
+class AProfileQirGenerator(AbsQirGenerator):
```
Comment (cqc-alec) [inline pytket/qir/conversion/profileqirgenerator.py]: A for "adaptive"? I think using the full name would be better -- and also changing the filename to reflect this.

Comment (cqc-melf) [reply pytket/qir/conversion/profileqirgenerator.py]: Yes, it is for Adaptive, I have expended this and added some more details to the docstring of the class.

File: pytket/qir/conversion/pytketqirgenerator.py (hunk 3 above)
Comment (cqc-melf) [reply pytket/qir/conversion/pytketqirgenerator.py]: I think this is not really helpful at the moment, I have remove this and added more details to the docs strings of the classes in the files. Maybe that is better?

File: pytket/qir/conversion/qirgenerator.py (hunk 2 above)
Comment (cqc-melf) [reply pytket/qir/conversion/qirgenerator.py]: I thought of `abstract` which is have now written fully. Open to other suggestions.

File: pytket/qir/conversion/api.py (hunk 1 above)
Comment (cqc-melf) [reply pytket/qir/conversion/api.py]: done

File: pytket/qir/conversion/pytketqirgenerator.py [hunk 5]
```diff
@@ -32,12 +27,17 @@
...
-    AbsQirGenerator,
+    AbstractQirGenerator,
 )
 
 
-class PytketQirGenerator(AbsQirGenerator):
-    """Generate QIR from a pytket circuit."""
+class PytketQirGenerator(AbstractQirGenerator):
+    """Generates QIR from a pytket circuit in line with the pytket profile.
+    This profiles uses the functions `get_creg_bit`, `set_creg_bit`,
```
Comment (cqc-alec) [inline pytket/qir/conversion/pytketqirgenerator.py]:
```suggestion
    This profile uses the functions `get_creg_bit`, `set_creg_bit`,
```

File: pytket/qir/conversion/qirgenerator.py [hunk 6]
```diff
@@ -112,8 +107,9 @@
 }
 
 
-class AbsQirGenerator:
-    """Generate QIR from a pytket circuit."""
+class AbstractQirGenerator:
+    """Abstract Class for the QIR generation from a pytket circuit.
+    Implementing the not profile specific functionality"""
```
Comment (cqc-alec) [inline pytket/qir/conversion/qirgenerator.py]:
```suggestion
    Implementing the functionality that is not specific to any profile"""
```

**Review by cqc-alec — APPROVED** [review]

**Pull request merged**